src/wasm/wasm-binary.cpp

```diff
@@ -16,68 +16,52 @@
 
 #include <algorithm>
 #include <fstream>
-#include <shared_mutex>
 
+#include "ir/module-utils.h"
 #include "support/bits.h"
 #include "wasm-binary.h"
 #include "wasm-stack.h"
 
 namespace wasm {
 
 void WasmBinaryWriter::prepare() {
-  // Collect function types and their frequencies
-  using Counts = std::unordered_map<Signature, size_t>;
-  using AtomicCounts = std::unordered_map<Signature, std::atomic_size_t>;
+  // Collect function types and their frequencies. Collect information in each
+  // function in parallel, then merge.
+  typedef std::unordered_map<Signature, size_t> Counts;
+  ModuleUtils::ParallelFunctionAnalysis<Counts> analysis(
+    *wasm, [&](Function* func, Counts& counts) {
+      if (func->imported()) {
+        return;
+      }
+      struct TypeCounter : PostWalker<TypeCounter> {
+        Module& wasm;
+        Counts& counts;
+
+        TypeCounter(Module& wasm, Counts& counts)
+          : wasm(wasm), counts(counts) {}
+
+        void visitCallIndirect(CallIndirect* curr) {
+          auto* type = wasm.getFunctionType(curr->fullType);
+          Signature sig(Type(type->params), type->result);
+          counts[sig]++;
+        }
+      };
+      TypeCounter(*wasm, counts).walk(func->body);
+    });
+  // Collect all the counts.
   Counts counts;
   for (auto& curr : wasm->functions) {
     counts[Signature(Type(curr->params), curr->result)]++;
   }
   for (auto& curr : wasm->events) {
     counts[curr->sig]++;
   }
-
-  // Parallelize collection of call_indirect type counts
-  struct TypeCounter : WalkerPass<PostWalker<TypeCounter>> {
-    AtomicCounts& counts;
-    std::shared_timed_mutex& mutex;
-    TypeCounter(AtomicCounts& counts, std::shared_timed_mutex& mutex)
-      : counts(counts), mutex(mutex) {}
-    bool isFunctionParallel() override { return true; }
-    bool modifiesBinaryenIR() override { return false; }
```

Comment:
Is it important to keep `modifiesBinaryenIR` somewhere?

Reply:
Oh, yikes, that's a very serious issue that I hadn't realized, and I think it affects other stuff too! Worse, we don't actually test it properly... I'll think about a general fix here, but it should probably be separate from this PR.

Reply:
Ok, fixed + added a test here, but I still want to find a more general solution outside of this PR...